File: src/test/scala/de/ust/skill/generator/c/GeneratorTest.scala
/*  ___ _  ___ _ _                                                            *\
** / __| |/ (_) | |       The SKilL Generator                                 **
** \__ \ ' <| | | |__     (c) 2013-18 University of Stuttgart                 **
** |___/_|\_\_|_|____|    see LICENSE                                         **
\*                                                                            */
package de.ust.skill.generator.c

import java.io.File

import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.junit.JUnitRunner

import de.ust.skill.main.CommandLine

/**
 * Scala specific tests.
 *
 * @author Timm Felden
 */
@RunWith(classOf[JUnitRunner])
class GeneratorTest extends FunSuite {

  private val sourceRootFolder = new File("src/test/resources/c/");

  private def deleteFiles(target : File) {
    if (target.isDirectory) {
      target.listFiles().foreach(deleteFiles)
    }
    target.delete()
  }

  def check(prefix : String, src : String, target : String) {
    // delete seems to cause sporadic test failure
    //    deleteFiles(new File(s"testsuites/c/src/generated/$target"))
    CommandLine.exit = { s ⇒ fail(s) }
    CommandLine.main(Array[String]("src/test/resources/c/" + src,
      "--debug-header",
      "-L", "C",
      "-p", prefix,
      "-o", "testsuites/c/src/generated/" + target))
  }

  test("annotation")(check("", "annotation.skill", "annotation"))

  test("date")(check("", "date.skill", "date"))
  test("basic types")(check("", "basic_types.skill", "basic_types"))
  test("empty")(check("", "empty.skill", "empty"))
  test("const")(check("", "const.skill", "const"))
  test("auto")(check("", "auto.skill", "auto"))
  test("float")(check("", "float.skill", "float"))
  test("node")(check("", "node.skill", "node"))
  test("subtypes")(check("", "subtypesExample.skill", "subtypes"))
  ignore("container 1")(check("", "container1.skill", "container1"))
  ignore("container user types")(check("", "container_user_type.skill", "container_user_type"))
  ignore("container annotation")(check("", "container_annotation.skill", "container_annotation"))
  ignore("container string")(check("", "container_string.skill", "container_string"))

  // Generate two bindings with prefix into the same directory, so that the test can compile two bindings together
  test("prefix sub")(check("sub", "subtypesExample.skill", "subtypes_prefix"))
  test("prefix date")(check("date", "date.skill", "date_prefix"))
}

File: src/test/java/de/ust/skill/generator/c/GeneratorTest.java
package de.ust.skill.generator.c;
import java.io.File;

import org.junit.BeforeClass;
import org.junit.Test;


public class GeneratorTest {

    private final static File sourceRootFolder = new File ( "src/test/resources/c/" );
    private final static File targetRootFolder = new File ( "testsuites/cTestSuite/src/generated" );

    @BeforeClass
    public static void prepare () {
        targetRootFolder.mkdirs ();
        deleteFolderContents ( targetRootFolder );
    }

    private void generateBinding ( String prefix, String sourceFile, String targetDir ) throws Exception {
        Main.main ( new String[]{"-p", prefix, sourceRootFolder.getAbsolutePath () + "/" + sourceFile, targetRootFolder.getAbsolutePath () + "/" + targetDir} );
    }

    @Test
    public void doGenerateBindings () throws Exception {
        generateBinding ( "", "date.skill", "date" );
        generateBinding ( "", "basic_types.skill", "basic_types" );
        generateBinding ( "", "empty.skill", "empty" );
        generateBinding ( "", "const.skill", "const" );
        generateBinding ( "", "auto.skill", "auto" );
        generateBinding ( "", "float.skill", "float" );
        generateBinding ( "", "node.skill", "node" );
        generateBinding ( "", "subtypesExample.skill", "subtypes" );
        generateBinding ( "", "container1.skill", "container1" );
        generateBinding ( "", "container_user_type.skill", "container_user_type" );
        generateBinding ( "", "container_annotation.skill", "container_annotation" );
        generateBinding ( "", "container_string.skill", "container_string" );
        generateBinding ( "", "annotation.skill", "annotation" );
        
        // Generate two bindings with prefix into the same directory, so that the test can compile two bindings together
        generateBinding ( "sub", "subtypesExample.skill", "subtypes_prefix" );
        generateBinding ( "date", "date.skill", "date_prefix" );
        
    }

    private static void deleteFolderContents ( File folder ) {
        File[] files = folder.listFiles ();
        if ( files != null ) {
            for ( File currentFile: files ) {
                if ( currentFile.isDirectory () ) {
                    deleteFolderContents ( currentFile );
                } else {
                    currentFile.delete ();
                }
            }
        }
        folder.delete ();
    }
}
